### PR TITLE
Function to copy a dataset to a new history

### PR DESCRIPTION
**Pull request by simonbray**
@mtekman, think this is PR-ready.

Fixes https://github.com/galaxyproject/bioblend/issues/280.

### CODE REVIEW

File: bioblend/galaxy/histories/__init__.py [hunk 1]
```diff
@@ -660,3 +660,29 @@ def download_history(self, history_id, jeha_id, outf,
...
+        """
+
+        payload = {
+            'content': dataset_id,
+            'source': hda_ldda,
+            'type': 'dataset'
+        }
+
+        url = self.gi._make_url(self, history_id, contents=True)
+        return self._post(payload=payload, url=url)
```
Comment (simonbray) [inline bioblend/galaxy/histories/__init__.py]:
```suggestion
        self._post(payload=payload, url=url)
```

Comment (mtekman) [reply bioblend/galaxy/histories/__init__.py]: I think return it, otherwise we won't have any idea from the shell if it was successful

Comment (simonbray) [reply bioblend/galaxy/histories/__init__.py]: Sure

**Comment by simonbray** [comment]
@nsoranzo, can you cast an eye over this when you have time please?

File: bioblend/galaxy/histories/__init__.py [hunk 2]
```diff
@@ -660,3 +660,29 @@ def download_history(self, history_id, jeha_id, outf,
...
+        Copy a dataset to a different history.
+
+        :type history_id: str
+        :param history_id: history ID to which the dataset should be copied
+
+        :type dataset_id: str
+        :param dataset_id: dataset ID
+
+        :type hda_ldda: str
+        :param hda_ldda: Source of the dataset to be copied: 'hda' (the default) or 'library'
```
Comment (nsoranzo) [inline bioblend/galaxy/histories/__init__.py]: I think this can also be `library_folder`.

File: bioblend/galaxy/histories/__init__.py [hunk 3]
```diff
@@ -660,3 +660,29 @@ def download_history(self, history_id, jeha_id, outf,
...
+
+        :rtype: dict
+        :return: Information about the dataset
```
Comment (nsoranzo) [inline bioblend/galaxy/histories/__init__.py]:
```suggestion
        :return: Information about the copied dataset
```

File: bioblend/galaxy/histories/__init__.py [hunk 4]
```diff
@@ -660,3 +660,29 @@ def download_history(self, history_id, jeha_id, outf,
         r.raise_for_status()
         for chunk in r.iter_content(chunk_size):
             outf.write(chunk)
+
+    def copy_dataset(self, history_id, dataset_id, source='hda'):
+        """
+        Copy a dataset to a different history.
```
Comment (nsoranzo) [inline bioblend/galaxy/histories/__init__.py]:
```suggestion
        Copy a dataset to a history.
```

File: bioblend/_tests/TestGalaxyHistories.py [hunk 5]
```diff
@@ -183,5 +183,13 @@ def test_download_history(self):
         finally:
             shutil.rmtree(tempdir)
 
+    def test_copy_dataset(self):
+        history_id = self.history["id"]
+        dataset1_id = self._test_dataset(history_id)
+        self.history_id2 = self.gi.histories.create_history('TestCopyDataset')['id']
+        self.gi.histories.copy_dataset(self.history_id2, dataset1_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyHistories.py]:
```suggestion
        copied_dataset = self.gi.histories.copy_dataset(self.history_id2, dataset1_id)
```

File: bioblend/_tests/TestGalaxyHistories.py [hunk 6]
```diff
@@ -183,5 +183,13 @@ def test_download_history(self):
...
+        self.assertFalse(self.gi.histories.show_history(self.history_id2)['empty'])
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyHistories.py]:
```suggestion
        self._wait_and_verify_dataset(copied_dataset['id'], expected_contents)
```

File: bioblend/_tests/TestGalaxyHistories.py [hunk 7]
```diff
@@ -183,5 +183,13 @@ def test_download_history(self):
         finally:
             shutil.rmtree(tempdir)
 
+    def test_copy_dataset(self):
+        history_id = self.history["id"]
+        dataset1_id = self._test_dataset(history_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyHistories.py]:
```suggestion
        expected_contents = "1\t2\t3"
        dataset1_id = self._test_dataset(history_id, contents=expected_contents)
```

**Comment by nsoranzo** [comment]
Thanks @simonbray!

**Pull request merged**